### PR TITLE
[MNG-7107] relax profile id validation, different from coordinate id

### PR DESCRIPTION
**Pull request by hboutemy**
see https://issues.apache.org/jira/browse/MNG-7107

### CODE REVIEW

**Review by michael-o — COMMENTED** [review]
What happened to `?+`, `?-`, and `?' only testing? We don't consider the question mark to be a starting char for the profile id.

**Comment by Tibor17** [comment]
As I have understood one old feature request that this pattern `-my-profile` would exclude the profile `my-profile`.
So I understood that the minus is part of the expression. Similar to the old expression `!my-profile`.
If `-` and `!` are the reserved characters, then other characters can be used, IMHO. WDYT?
Of course the limitation might be imposed to the first character and not the other, in this case. WDYT?

**Comment by Tibor17** [comment]
I think this implementation covers my requirement mentioned in Jira.
Go ahead with the commit.
Feel free to add a test into it.

**Review by Tibor17 — APPROVED** [review]

**Comment by michael-o** [comment]
I would still ike to have clarified for `?, `-`, `+`, and `!`.

**Comment by Tibor17** [comment]
@michael-o 
You mean something like this?
`-P ?+my-profile` or `-P +my-profile`, is it an expression? If it is, what kind of meaning?

**Comment by michael-o** [comment]
> 
> 
> @michael-o
> You mean something like this?
> `-P ?+my-profile` or `-P +my-profile`, is it an expression? If it is, what kind of meaning?

Yes and both are valid iinput. In both cases the profile name is `my-profile`.

**Comment by hboutemy** [comment]
I coded the few cases I thought I knew about `-P` in `mvn` command line
but it seems I don't know what `?` means: I don't see it in code https://github.com/apache/maven/blob/master/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java#L1475
and I missed `!`, that IIUC seems to be a synonym to `-`

I'll add `!` as excluded first character, but for `?`, I'll need some explanations as why we should not accept that char

**Comment by michael-o** [comment]
A recent change make profile activatoins mandatory, unless it is prefixed with `?`. See https://github.com/apache/maven/commit/8defd169651db30fe0ac3ad3f318e471f0241d02.

**Comment by hboutemy** [comment]
ok, I found "optional profile activation" https://github.com/apache/maven/blob/master/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java#L1515

IMHO, this will require some reference documentation in https://maven.apache.org/ref/current/maven-embedder/

**Comment by michael-o** [comment]
I will test your change later this day...

**Comment by Tibor17** [comment]
It would be worth for future maintenance to create a new class checking these characters in profile id because both classes `DefaultModelValidator` and `MavenCli` are chcecking the first (and second character `?`) character, so it would worth to change the algorithm in one place instead of two.

**Review by mthmulders — APPROVED** [review]

File: maven-model-builder/src/main/java/org/apache/maven/model/validation/DefaultModelValidator.java [hunk 1]
```diff
@@ -916,7 +958,7 @@ private boolean isValidIdWithWildCards( String id )
 
     private boolean isValidIdWithWildCardCharacter( char c )
```
Comment (rfscholte) [inline maven-model-builder/src/main/java/org/apache/maven/model/validation/DefaultModelValidator.java:959]: This methodname is now weird: we're going from an general `isValidIdWithWildCardCharacter` to a more explicit `isValidCoordinateIdCharacter`.

File: maven-model-builder/src/main/java/org/apache/maven/model/validation/DefaultModelValidator.java [hunk 2]
```diff
@@ -850,36 +853,75 @@ private boolean validateId( String prefix, String fieldName, ModelProblemCollect
...
+            }
+            validProfileIds.add( id );
+            return true;
+        }
+    }
+
+    private boolean isValidProfileId( String id )
+    {
+        switch ( id.charAt( 0 ) )
+        { // avoid first character that has special CLI meaning in "mvn -P xxx"
```
Comment (rfscholte) [inline maven-model-builder/src/main/java/org/apache/maven/model/validation/DefaultModelValidator.java:914]: IMO a profileId should never contain a comma or a space

Comment (hboutemy) [reply maven-model-builder/src/main/java/org/apache/maven/model/validation/DefaultModelValidator.java:914]: yes, there are good ideas during the discussion on improvements for profile and coordinate id checks: I'll let that as improvements "up for grabs": I needed early fix to stop failing on "jdk9+" profile id
Now I'll let improvement ideas go at their own pace

File: maven-model-builder/src/main/java/org/apache/maven/model/validation/DefaultModelValidator.java (hunk 1 above)
Comment (hboutemy) [reply maven-model-builder/src/main/java/org/apache/maven/model/validation/DefaultModelValidator.java:959]: ok, I'll rename this method also

**Comment by hboutemy** [comment]
bq. It would be worth for future maintenance to create a new class checking these characters in profile id because both classes DefaultModelValidator and MavenCli are chcecking the first (and second character ?) character, so it would worth to change the algorithm in one place instead of two.

I understand the general idea, but IMHO, this would create more complexity (with dependency tree between artifacts, and abstraction) than real benefit.

But don't hesitate to work on this later if you really think there is a benefit

**Comment by Tibor17** [comment]
There does not need to be an abstraction. Two methods in one class which means that the developer is not able to change the first method if modifying another one.

File: maven-model-builder/src/main/java/org/apache/maven/model/validation/DefaultModelValidator.java (hunk 2 above)
Comment (michael-o) [reply maven-model-builder/src/main/java/org/apache/maven/model/validation/DefaultModelValidator.java:916]: I don't understand the urgent need for the `jdk9+`. This is alpha code, why did you put this into production?

**Comment by michael-o** [comment]
@hboutemy Why didn't you wait for me to test? This one is incomplete. It misses a handling for these:
```
diff --git a/maven-model-builder/src/test/resources/poms/validation/invalid-profile-ids.xml b/maven-model-builder/src/test/resources/poms/validation/invalid-profile-ids.xml
index 74b670b9d..7419fd83d 100644
--- a/maven-model-builder/src/test/resources/poms/validation/invalid-profile-ids.xml
+++ b/maven-model-builder/src/test/resources/poms/validation/invalid-profile-ids.xml
@@ -37,6 +37,21 @@ under the License.
         <profile>
             <id>?invalid-id</id>
         </profile>
+        <profile>
+            <id>-?invalid-id</id>
+        </profile>
+        <profile>
+            <id>+?invalid-id</id>
+        </profile>
+        <profile>
+            <id>!?invalid-id</id>
+        </profile>
+        <profile>
+            <id>invalid,id</id>
+        </profile>
+        <profile>
+            <id>invalid id</id>
+        </profile>
         <profile>
             <id>valid-id</id>
         </profile>
```

**Comment by michael-o** [comment]
There likely more chars which would be forbidded, e.g., <, >, |, ~, ;, $, %, ", ', [, ], *, \. They all have some meaning either in POSIX shell, bash, PowerShell or Windows cmd. Likely a whilelist after the first/second char would make more sense!

**Comment by hboutemy** [comment]
@Tibor17 don't hesitate to implement, I'm eager to see the relationship between maven-model-builder and maven-embedder

@michael-o sorry, I had already much feedback, and my intent was not to implement any and all validations but just relax a too hard check that was in master and made current HEAD unusable
don't hesitate to add even more checks if you want in a further issue: I'm not against by don't have time for that

**Comment by jira-importer** [comment]
Resolve #7965

**Pull request merged**